### PR TITLE
add pageserver component-level benchmark (`pagebench`)

### PR DESCRIPTION
**Pull request by problame**
This PR adds a component-level benchmarking utility for pageserver.
Its name is `pagebench`.

The problem solved by `pagebench` is that we want to put Pageserver under high load.

This isn't easily achieved with `pgbench` because it needs to go through a compute, which has signficant performance overhead compared to accessing Pageserver directly.

Further, compute has its own performance optimizations (most importantly: caches). Instead of designing a compute-facing workload that defeats those internal optimizations, `pagebench` simply bypasses them by accessing pageserver directly.

Supported benchmarks:

* getpage@latest_lsn
* basebackup
* triggering logical size calculation

This code has no automated users yet.
A performance regression test for getpage@latest_lsn will be added in a later PR.

part of https://github.com/neondatabase/neon/issues/5771

### CODE REVIEW

**Comment by hlinnaka** [comment]
Cool! How do you use it?

**Comment by problame** [comment]
`./target/debug/pagebench`

**Review by jcsp — APPROVED** [review]
I started trying this out and immediately got distracted with profiling :-)  Super useful tool.  Minor comments only.

**Comment by jcsp** [comment]
Noticed when testing this that PageStreamClient::getpage always sets `latest=False` -- this causes the page server to skip caching relation sizes in Timeline.  It's kind of interesting  to see the impact of that (reads that aren't to latest LSN are rather expensive with the extra getpage for relation size), but probably not what was intended

**Comment by jcsp** [comment]
Tracking the glitches from testing here (don't need to block this PR):
- https://github.com/neondatabase/neon/issues/6209
- https://github.com/neondatabase/neon/issues/6210

File: pageserver/pagebench/src/cli/targets.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+use crate::util::{
+    discover_timelines::get_pageserver_tenant_timelines, tenant_timeline_id::TenantTimelineId,
+};
+
+pub(crate) struct Spec {
+    pub(crate) limit_to_first_n_targets: Option<usize>,
+    pub(crate) targets: Option<Vec<TenantTimelineId>>,
+}
+
+pub(crate) async fn discover(
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/pagebench/src/cli/targets.rs:15]: The idea is that stuff in the `cli` module subtree is specific to the benchmarking CLI.

I guess I could move `discover_timelines.rs` into some `utils` / `helpers` ext trait in the mgmt API crate.
WDYT?

File: pageserver/pagebench/src/util/tokio_thread_local_stats.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            );
+        }
+    };
+}
+
+use std::sync::{Arc, Mutex};
+
+pub(crate) use declare;
+
+macro_rules! main {
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/pagebench/src/util/tokio_thread_local_stats.rs:18]: I'm all for more verbose, but, often find it hard to convince others.

This is the commit that tried to factor out thread local stats: https://github.com/neondatabase/neon/pull/6174/commits/c417a23dd0af81c8d60088f21bce44fae68956eb

You prefer to revert it?

[the remark this answers is not in the document]
Comment (problame) [reply pageserver/pagebench/src/util/tokio_thread_local_stats.rs:18]: Hm, I think I still believe the macro is a slight win. 

Let's see how many additional benchmark subcommands we evolve and what the need for sharing will be.

**Pull request merged**